File: 08-2.py
#!/usr/bin/env python

import re
from collections import namedtuple
from math import lcm

lines = [row.strip() for row in open('08.input').readlines()]

line_format = re.compile("^(?P<from>[0-9A-Z]+) = \\((?P<left>[0-9A-Z]+), (?P<right>[0-9A-Z]+)\\)$")

insns = lines[0]

Node = namedtuple('Node', ['L', 'R'])
nodes = {}
for line in lines[2:]:
    m = line_format.match(line)
    nodes[m['from']] = Node(L=m['left'], R=m['right'])

start_nodes = [n for n in nodes.keys() if n.endswith('A')]
num_threads = len(start_nodes)
steps = [0] * num_threads

for n in range(num_threads):
    ip = 0
    node = start_nodes[n]
    while not node.endswith('Z'):
        node = getattr(nodes[node], insns[ip])
        ip = (ip + 1) % len(insns)
        steps[n] += 1

print(lcm(*steps))

File: 08-2.rb
#!/usr/bin/env ruby

require 'ostruct'

lines = File.read('08.input').lines.map(&:strip)

line_format = /^(?<from>[[:alnum:]]+) = \((?<left>[[:alnum:]]+), (?<right>[[:alnum:]]+)\)$/

insns = lines[0].chars.map do |i|
  if i == 'L'
    'left'
  else
    'right'
  end
end

nodes = {}

lines[2..].each do |line|
  m = line.match line_format
  nodes[m['from']] = OpenStruct.new({ left: m['left'], right: m['right'] })
end

start_nodes = nodes.keys.select { |k| k.end_with? 'A' }
num_threads = start_nodes.size
steps = [0] * num_threads

num_threads.times do |n|
  ip = 0
  node = start_nodes[n]
  until node.end_with? 'Z'
    node = nodes[node][insns[ip]]
    ip = (ip + 1) % insns.size
    steps[n] += 1
  end
end

print steps.reduce(1, :lcm), "\n"
